source/common/common/shared_memory_hash_set.h

```diff
@@ -0,0 +1,370 @@
+      return false;
+    }
+
+    if (control_->size > options_.capacity) {
+      ENVOY_LOG(error, "SharedMemoryHashSet size={} > capacity={}", control_->size,
+                options_.capacity);
+      return false;
+    }
+
+    // As a sanity check, make sure there are control_->size values
+    // reachable from the slots, each of which has a valid char_offset
+    uint32_t num_values = 0;
+    for (uint32_t slot = 0; slot < options_.num_slots; ++slot) {
+      uint32_t next = 0; // initialized to silence compilers.
+      for (uint32_t cell_index = slots_[slot]; cell_index != Sentinal; cell_index = next) {
+        if (cell_index >= options_.capacity) {
+          ENVOY_LOG(error, "SharedMemoryHashSet cell index too high={}, capacity={}", cell_index,
+                    options_.capacity);
+          ret = false;
+          break;
+        } else {
+          Cell& cell = getCell(cell_index);
+          absl::string_view key = cell.value.key();
+          if (computeSlot(key) != slot) {
+            ENVOY_LOG(error, "SharedMemoryHashSet hash mismatch for key={}", std::string(key));
+            ret = false;
+          }
+          next = cell.next_cell;
+          ++num_values;
+          // Avoid infinite loops if there is a next_cell cycle within
+          // a slot. Note that the num_values message will be emitted
+          // outside the loop.
+          if (num_values > control_->size) {
+            break;
+          }
+        }
+      }
+    }
+    if (num_values != control_->size) {
+      ENVOY_LOG(error, "SharedMemoryHashSet has wrong number of live cells: {}, expected {}",
+                num_values, control_->size);
+      ret = false;
+    }
+    return ret;
+  }
+
+  /**
+   * Returns a string describing the contents of the map, including the control
+   * bits and the keys in each slot.
+   */
+  std::string toString() {
+    std::string ret;
+    ret = fmt::format("options={}\ncontrol={}\n", options_.toString(), control_->toString());
+    for (uint32_t i = 0; i < options_.num_slots; ++i) {
+      ret += fmt::format("slot {}:", i);
+      for (uint32_t j = slots_[i]; j != Sentinal; j = getCell(j).next_cell) {
+        ret += " " + std::string(getCell(j).value.key());
+      }
+      ret += "\n";
+    }
+    return ret;
+  }
+
+  /**
+   * Initializes a hash-map on raw memory. No expectations are made about the state of the memory
+   * coming in.
+   * @param memory
+   */
+  void init(uint8_t* memory, uint32_t num_bytes) {
+    mapMemorySegments(memory);
+    RELEASE_ASSERT(num_bytes == numBytes(options_));
+    control_->hash_signature = HashUtil::xxHash64(signatureStringToHash());
+    control_->num_bytes = num_bytes;
+    control_->options = options_;
+    control_->size = 0;
+    control_->free_cell_index = 0;
+
+    // Initialize all the slots;
+    for (uint32_t slot = 0; slot < options_.num_slots; ++slot) {
+      slots_[slot] = Sentinal;
+    }
+
+    // Initialize all the key-char offsets.
```

Comment:
I'm not sure what this comment means.  Maybe change to "Initialize the free-cell list."?